ENYO-3309: Added the new Notification Component

Djspaceg opened the pull request.

Commit:
Added a new component: Notification

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
Don't leave in, that which you don't use...

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
Updated Notification to reflect the latest GUI design.

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

webOS101 commented on css/moonstone-variables.less:
```diff
@@ -93,6 +93,7 @@
 @moon-expandable-value-font-size:   27px;
 @moon-expandable-client-item-font-size:   27px;
 @moon-divider-font-size:            24px;
+@moon-notification-font-size:       30px;
```
This value appears to be duplicated below

Djspaceg replied:
Got it. Good catch.

webOS101 commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,417 @@
+require('moonstone');
+
+/**
+* Contains the declaration for the {@link module:moonstone/Popup~Popup} kind.
```
This should be Notification, not Popup

webOS101 commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,417 @@
+	/**
+	* @private
+	*/
+	// floating: true,
```
Delete this?

webOS101 commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,417 @@
+	*/
+	handlers: {
+		onRequestScrollIntoView   : '_preventEventBubble',
+		// ontransitionend           : 'animationEnd',
```
remove

Commit:
Made this inherit from Popup to add some necessary features.

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

webOS101 commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,433 @@
+* from the bottom of the screen.
+*
+* @class Notification
+* @ui
```
Can you add an `@extends module:enyo/Popup~Popup`?

webOS101 commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,433 @@
+		content: '',
+
+		/**
+		* When `true`, the notification will animate on/off screen.
```
I think this description needs an update.

webOS101 commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,433 @@
+		* @default true
+		* @public
+		*/
+		spotlightModal: true,
```
I don't think this is hooked up to anything yet, is it?

Djspaceg replied:
It's evidently implemented in Spotlight. It has no visible wiring in both enyo/Popup and moonstone/Popup.

aarontam replied:
As a general note and/or FYI, `spotlightDisabled` has a similar behavior of only being referenced in `Spotlight`.

Commit:
Oh Travis.
My linter is apparently not working any more...

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
Removed duplicate variable and cleaned up docs a little

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
Look at all this unnecessary stuff! Also incrementally improved docs

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
More oops. This fixes pointer activation with 5-way next action.

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
Who wrote this code? So sloppy.

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

Commit:
Duplicated line

Enyo-DCO-1.1-Signed-off-by: Blake Stephens <blake.stephens@lge.com>

aarontam commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,351 @@
+	/**
+	* @private
+	*/
+	release: function () {
```
Think we can get rid of this, as the implementation is the same as the base kind's.

aarontam commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,351 @@
+
+	/**
+	* @private
+	* @lends module:moonstone/Popup~Popup.prototype
```
Should reference `Notification` here, instead. Also, we were moving away from published properties - can we just declaring these as properties directly on the kind?

Djspaceg replied:
I was initially under the impression we were deprecating `published` but saw no evidence of it, outside my own work. I used it here because it seemed like nobody was deeply interested in not having it any more. I can take it out, but it will just make this control look different from others. Thoughts?

aarontam commented on src/Notification/Notification.js:
```diff
@@ -0,0 +1,351 @@
+	hidden: function (sender, ev) {
+		this.respotActivator();
+
+		if (this._initialized && this.allowBackKey && !EnyoHistory.isProcessing()) {
```
What do you think about checking for something like `this.hasNode()` here, instead of maintaining an internal property? Would also allow us to get rid of the `render` override.

Djspaceg replied:
This code was actually copied directly from moonstone/Popup, so we should ideally update that too. I'll swap out `_initialized` for `this.hasNode()` or `this.generated` or something to that same effect.

lishammel commented:
Closing PR as the jira has been closed.
